test/sqllogictest/transform/join_nonnull_requirements.slt

```diff
@@ -0,0 +1,123 @@
+# Copyright Materialize, Inc. AND contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+mode cockroach
+
+statement ok
+CREATE TABLE t1 (
+    f1 int NOT NULL,
+    f2 int
+)
+
+# non-null requirement on f2 derived from the join equality
+query T multiline
+EXPLAIN SELECT * FROM t1 AS a1, t1 AS a2 WHERE a1.f1 = a2.f2;
+----
+%0 =
+| Get materialize.public.t1 (u1)
+| ArrangeBy (#0)
+
+%1 =
+| Get materialize.public.t1 (u1)
+| Filter !(isnull(#1))
+
+%2 =
+| Join %0 %1 (= #0 #3)
+| | implementation = Differential %1 %0.(#0)
+| | demand = (#0..#2)
+| Project (#0..#2, #0)
+
+EOF
+
+# non-null requirement on f2 derived from the join equality
+query T multiline
+EXPLAIN SELECT * FROM t1 AS a1, t1 AS a2 WHERE a1.f1 = a2.f2 + 1;
```

Comment:
This doesn't exercise the new code, since the non-null requirement is extracted from the filter before it becomes a join equivalence. The in join_fusion.slt does exercise the new code though.